### ISSUE TITLE
Enforce HTTPS: More details for scenarios when app doesn't accept HTTPS (proxy/load balancer)

### ISSUE THREAD

**Issue by justinhelgerson** [aspnet/AspNetCore.Docs#10926]
I think it would be helpful if the docs on enforcing HTTPS more explicitly called out that you can't use the HTTPS redirection middleware (`UseHttpsRedirection`) in a reverse proxy scenario **unless your application is accepting HTTPS requests**.

I could be mistaken, but I suspect that a lot of people don't have their applications setup to accept HTTPS requests in a load balanced environment.  Instead, the common workflow is that the load balancer is responsible for off loading or terminating the HTTPS request and sending the HTTP request to the application.  `UseHttpsRedirection` forces you into having your application listen for HTTPS requests (so the redirect middleware can work) even though you would never receive an HTTPS request at the app level.

Related question on Stack Overflow:  https://stackoverflow.com/questions/54674622/redirecting-http-to-https-behind-load-balancer

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: c7ddfb2f-5dea-5e31-2a27-02a4dc16104c
* Version Independent ID: 12ca0b9a-2638-a4ce-c45f-d4579c83dafb
* Content: [Enforce HTTPS in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/security/enforcing-ssl?view=aspnetcore-2.2&tabs=visual-studio#port-configuration)
* Content Source: [aspnetcore/security/enforcing-ssl.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/security/enforcing-ssl.md)
* Product: **aspnet-core**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by Rick-Anderson** [comment]
> `UseHttpsRedirection`) in a reverse proxy scenario requires app to accept HTTPS

That's what I would expect.

> a lot of people don't have their applications setup to accept HTTPS requests in a load balanced environment.

That's not my understanding.

@blowdart should we update this page with these comments?

**Comment by blowdart** [comment]
In a proxy situation you should be using the forwarded headers middleware which would then cause the inbound protocol to be whatever it was when you hit the proxy. The HTTPS enforcement middleware would be doing the right thing.

So no, this change shouldn't be made at all. It isn't a misunderstanding with the HTTPS enforcement, it's a misunderstanding around not using the headers middleware.

**Comment by blowdart** [comment]
@Rick-Anderson The answer here is I don't know. Let Justin figure out if there's a better way.

**Comment by blowdart** [comment]
`HttpsRedirectionOptions.HttpsPort`, setting the ASPNETCORE_HTTPS_PORT environment variable, or by setting an HTTPS url in launchsettings.json.

Set the port to 443 in options, and then, in conjunction with a properly configured forwarding headers middleware that should get the right behaviour maybe?

**Comment by justinhelgerson** [comment]
@Rick-Anderson I would think SSL termination is quite common?  The change in the middleware here requires me to setup a "dummy" HTTPS listener on my app which ends up not being used except for the middleware to do detection.  It seems a bit odd to me, but I'm sure this makes sense for other folks.  For me it requires additional scripts & steps in our deploy process that I'm going to avoid.  

@blowdart That was my thought as well.  I've tried setting the options in conjunction with the middleware, but no dice.  I think the issue with the workflow is (simplified):

1. HTTP request is sent
1. Request hits load balancer
1. Request is sent to application *without* forward header because the originating request wasn't HTTPS
1. Middleware detects HTTP and determines if it can redirect, but no HTTPS port is detected
1. Request stays as HTTP and no redirection occurs

The workaround seems to be using the `AddRedirectToHttps` middleware.

**Comment by blowdart** [comment]
@jkotalik UseHttpsRedirection / AddRedirectToHttps. Why on earth are there two for starters? And why the restriction in UseHttpsRedirection?

**Comment by jkotalik** [comment]
Sorry I let this slip off my radar. I now started to follow the docs repo 😄 

> Oh wait, I just saw the change. @jkotalik the change to HttpsRedirectionMiddleware was bad in some cases, and ought to be reexamined, unless you think you can change it with headers.

@blowdart Which change are you referring to? The only one I can think of is that the HttpsRedirectionMiddleware will not force redirect to HTTPS, only warn.

> @jkotalik UseHttpsRedirection / AddRedirectToHttps. Why on earth are there two for starters? And why the restriction in UseHttpsRedirection?

AddRedirectoToHttps is part of the UrlRewrite Middleware, which was added in 1.1. During the 2.1 timeline, people wanted the HTTPS redirection to have it's own middleware for our work to make HTTPS the default. It's odd, I agree, but UrlRewrite requires the ability to redirect HTTPS and people wanted HTTPS in a new namespace.
